concordium-std/src/types.rs

```diff
@@ -1,11 +1,100 @@
-use crate::num::NonZeroU32;
+use crate::{num::NonZeroU32, HasContractStateEntry, HasContractStateLL};
+use std::{cell::RefCell, marker::PhantomData, rc::Rc};
+
+#[derive(Debug)]
+pub struct StateMap<K, V, S>
+where
+    S: HasContractStateLL, {
+    pub(crate) _marker_key:   PhantomData<K>,
+    pub(crate) _marker_value: PhantomData<V>,
+    pub(crate) prefix:        StateItemPrefix,
+    pub(crate) state_ll:      Rc<RefCell<S>>,
+}
+
+#[derive(Debug)]
+pub struct StateMapIter<K, V, S: HasContractStateLL> {
+    pub(crate) state_iter:    S::IterType,
+    pub(crate) state_ll:      Rc<RefCell<S>>,
+    pub(crate) _marker_key:   PhantomData<K>,
+    pub(crate) _marker_value: PhantomData<V>,
+}
+
+#[derive(Debug)]
+pub struct StateSet<T, S>
+where
+    S: HasContractStateLL, {
+    pub(crate) _marker:  PhantomData<T>,
+    pub(crate) prefix:   StateItemPrefix,
+    pub(crate) state_ll: Rc<RefCell<S>>,
+}
+
+#[derive(Debug)]
+pub struct StateSetIter<T, S: HasContractStateLL> {
+    pub(crate) state_iter: S::IterType,
+    pub(crate) state_ll:   Rc<RefCell<S>>,
+    pub(crate) _marker:    PhantomData<T>,
+}
+
+#[derive(Debug)]
+pub struct StateBox<T, S: std::fmt::Debug> {
+    pub(crate) prefix:   StateItemPrefix,
+    pub(crate) state_ll: Rc<RefCell<S>>,
+    pub(crate) _marker:  PhantomData<T>,
+}
+
+#[derive(Debug, Default)]
+pub struct ContractStateLL;
+
+pub struct ContractStateIter {
+    pub(crate) iterator_id: StateIteratorId,
+}
+
+pub type StateEntryId = u32;
+pub type StateIteratorId = u32;
+pub type StateItemPrefix = Vec<u8>;
 
-/// A type representing the constract state bytes.
 #[derive(Default)]
-pub struct ContractState {
+pub struct StateEntry {
+    pub(crate) state_entry_id:   StateEntryId,
     pub(crate) current_position: u32,
 }
 
+pub struct VacantEntryRaw<StateEntryType>
+where
+    StateEntryType: HasContractStateEntry, {
+    pub(crate) state_entry: StateEntryType,
+}
+
+pub struct OccupiedEntryRaw<StateEntryType>
+where
+    StateEntryType: HasContractStateEntry, {
```

Comment:
Do all of these constraints on the structs provide value?

In my experience they were usually not very useful and tended to clutter things, forcing the constraint even for methods that did not require them.

Reply:
I think they are left there by accident. I'll remove them.